src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs

```diff
@@ -457,51 +456,42 @@ private static void ValidateAccessor(Expression instance, MethodInfo method, Par
 
         private static void ValidateAccessorArgumentTypes(MethodInfo method, ParameterInfo[] indexes, ref ReadOnlyCollection<Expression> arguments)
         {
-            if (indexes.Length > 0)
+            if (indexes.Length != arguments.Count)
             {
-                if (indexes.Length != arguments.Count)
+                throw Error.IncorrectNumberOfMethodCallArguments(method);
+            }
+
```

Comment:
If we're no longer going to take the shorter path on `indexes.Length == 0` at the start, could we `return` in such a case?